### ISSUE TITLE
Fast way to pull in transcripts

### ISSUE THREAD

**Issue by ceedubs** [unisonweb/unison#3314]
# Problem

Often in transcripts it's useful to be able to `pull` (especially the base library; it's really hard to write a transcript without it). But running `ucm transcript` on a transcript that pulls means pulling the entire library again, which might take minutes to run. This is especially annoying when you realize a small typo in the transcript and run it again only to find that you are downloading the library again.

`transcript.fork` is one solution to this problem, but it forks the entirety of the codebase, so you risk name collisions. It's often useful to have an empty root namespace at the beginning of the transcript execution.

# Goal

Make a near-zero barrier for one Unison user to share a transcript with another and have it run quickly on their machine.

As a bonus, this would probably make it more ergonomic to run integration tests from transcripts.

# Current workarounds

There are a couple of things that you can currently do to mitigate this issue.

## Fork from a transcript-specific codebase

This is a trick that @aryairani taught me.

1. Extract the `pull` commands from your transcript into a separate `download-stuff.md` transcript.
2. Run `ucm transcript --save-codebase download-stuff.md`. The output should include something like `You can run `unison --codebase /tmp/.../transcript-9a10f15a4fd56753` to do more work with it.
3. Run `ucm --codebase /tmp/.../transcript-9a10f15a4fd56753 transcript.fork my-actual-transcript.md`

The end effect of this solution is what you want, but it's several extra steps for the developer, and it makes it so that you can't just share `my-actual-transcript.md` with someone and expect it to work for them.

## Only pull squashed releases

If you only ever pull squashed releases, the pulls are fairly fast, so this is less of an issue.

However, the pulls still aren't _that_ fast. And this isn't always an option, because sometimes you are trying to share a reproducible error that happens with some non-squashed code.

# Solution brainstorming

Since everything is content-addressed, it seems like Unison should be in a good place to solve this problem. What can we do?

Below are some possibilities that come to mind.

**NOTE**: some of these possibilities introduce behavior that might cover up issues that could arise in completely fresh codebases. We could potentially create a new `fork.fresh` to maintain current behavior for Unison/ucm tests. But I think that if we want transcripts to be ergonomic for end users, the current `transcript` command should change.

## `pull` without a provided `--codebase` can pull from the default codebase

If a transcript does `.> pull unison.public.base.latest`, then the transcript would hit Share to see what's there, but it would pull from the local default codebase instead of doing a remote pull for everything that is available in the local codebase.

This seems kind of ideal to me, but I don't know if we have support from pulling from one local codebase to another like this; it might be tricky to implement.

## fork types/terms from existing codebase, but not namespaces

If we had something in between the current `transcript` and `transcript.fork` where it copied types/term definitions from the existing codebase but _didn't_ copy any namespaces or history, would that dramatically improve the performance of doing a `pull` in the transcript? If so, this seems like a fairly easy solution; but I don't have a good sense of how much it would help.

## use transcript/temp namespace in existing codebase

What if transcripts acted on the existing codebase, but transparently worked within a temporary and/or transcript namespace? So instead of a separate `/tmp/.../transcript-9a10f15a4fd56753` codebase on your filesystem you get a `._transcripts.9a10f15a4fd56753` namespace in your existing codebase. Maybe these automatically clean themselves up, or maybe users just know that they can probably safely `delete.namespace ._transcripts` at any point.

## Other ideas?

People who work on `ucm` probably have better ideas than I do.

**Comment by ceedubs** [comment]
@ChrisPenner weighed in on Slack:

> A really easy fix for this would be to seed all transcript codebases with your main codebases content addressed definitions, but without any namespace structure. That way all the pulls would be instant since you'd have the data already it'd just need to add a name for it.
>
> This would bloat the size of transcript codebases, but they're almost always deleted immediately anyways.

In the off-chance that the transcript repo doesn't get deleted (for me it always has when I don't use the `--save-codebase` flag; often to my chagrin), it's put into a `tmp` dir, so I don't think that this is a concern.

He clarified that this is different than the "fork types/terms from existing codebase, but not namespaces" proposal:

> It would keep everything including namespaces and history, but would just set the root namespace to the empty namespace

If this solution is low-hanging fruit, then it sounds great to me! My vote would be to make this the default for `transcript`. For existing tests we could either create a new `transcript.fresh` (or whatever we want to call it) or possibly you could get the desired functionality with doing something like `--codebase-create $(mktemp -d)`.

**Comment by ChrisPenner** [comment]
> fork types/terms from existing codebase, but not namespaces

Yeah we could do a variant of this, where we clone the user's codebase via sqlite and just reset the root-namespace to an empty branch. That way the pulls are all instant if the user already has the code.

**Comment by aryairani** [comment]
I agree that it's increasingly unrealistic to write interesting transcripts without pulling base.

A few thoughts...
* A transcript that uses `pull` won't generally be reproducible, because the thing at the other end of the URL can change over time.  Only pulling a particular namespace hash seems reproducible-enough to me to be considered reproducible in the way we like to think of transcripts as being.
* If we had some command to obliterate history https://github.com/unisonweb/unison/issues/3316, you could use it to clear your root namespace. But you still have to transcript.fork which is weird when you don't want that namespace, and then wipe your namespace, and then still `pull` which as we said isn't generally reproducible anyway.
* You can use the `reset-root` command to obliterate history too, if you know the hash of the empty history, which you don't. I'm also not sure under what conditions it would be present in the codebase and succeed.

I like "pull without a provided --codebase can pull from the default codebase" above. Like.. it's _just_ a mode for running transcripts quickly, using another local codebase as a cache? `--use-cache <codebase>` maybe multiple times. Before you try to sync an entity over the network, you try to sync it from the cache(s).

**Comment by ceedubs** [comment]
Note: `pull.without-history` now is faster than it used to be. But it still would be nice to have a fast way to pull from a local codebase or execute in an isolated namespace or something.
